### DISCUSSION TITLE
How to make Microsoft Presidio work with Excel?

### DISCUSSION THREAD

**Discussion by Welding-Torch** [microsoft/presidio#1300] [Q&A] [answered]
In short: I have an Excel sheet with sensitive data. How do I scrub the sensitive data out of it and then share it with people? 

Bonus question: How can I use Microsoft Presidio with Google Sheets?

**Comment by omri374** [comment]
Hi, one option would be to read the excel using Pandas and pass it through the new presidio-structured package (if you'd like full cells to be redacted), or through the `BatchAnalyzerEngine` and `BatchAnonymizerEngine` if you'd like to scan each text in each cell and remove only the PII parts. With presidio-structured, a cell containing "My name is David" would be redacted completely, together with all cells in that column. With the batch analysis and anonymization mode, the cell will be replaced with "My name is <PERSON>".

References:
1. Pass a Pandas data frame (after reading the excel with `pandas.read_excel`): https://microsoft.github.io/presidio/structured/
2. Batch analysis/anonymization: https://github.com/microsoft/presidio/blob/main/docs/samples//python/batch_processing.ipynb

For Google Sheets, the easiest would be to download it to excel/csv and pass it to Presidio. One can think of an integration of Presidio into Google Sheets, but Presidio does not have that capability built into the tool.

**Reply by Welding-Torch** [reply]
If I want to give Microsoft Presidio as a simple program to a friend in office who doesn't know how to code but deals with Excel spreadsheets daily, how can I make it really, really simple for him to use?

Maybe like a one-click 'Anonymize all data' button?

I'd be super grateful if you could help me figure this out.

P.S: I don't want the data to be removed, I want it to be substituted for something anonymous. Eg: "My name is David" will become "My name is John".

**Reply by omri374** [reply]
I would look into something like [streamlit](https://streamlit.io/), which was used for [Presidio's demo website](https://huggingface.co/spaces/presidio/presidio_demo). It would allow you to package it as a standalone app.

You could look into tweaking our [demo website's code](https://github.com/microsoft/presidio/blob/main/docs/samples/python/streamlit/index.md) for your purposes.

Note that there's a tradeoff between the two options. The first one (presidio-structured) would assume everything in a column is PII if enough samples were found to contain PII. This would likely reduce false positives. The other one (batch analyzer / anonymizer) is tending towards utility in the privacy-utility tradeoff.

**Reply by Welding-Torch** [reply, marked as answer]
I built it: https://github.com/Welding-Torch/Anonymize_Excel

**Reply by omri374** [reply]
Cool!
